File: backend/src/main/java/com/resturant/management/rms/order/Order.java
package com.resturant.management.rms.order;

import com.resturant.management.rms.common.BaseAuditEntity;
import com.resturant.management.rms.dining.DiningTable;
import com.resturant.management.rms.user.UserInfm;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * A bill. Table {@code orders} — {@code order} is a reserved SQL word.
 *
 * <p>Totals are stored, not computed on read, so a reprinted receipt always
 * shows what was actually charged even if prices or the VAT rate change later.
 */
@Entity
@Table(name = "orders")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Order extends BaseAuditEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "invoice_no", nullable = false, unique = true, length = 20)
    private String invoiceNo;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "table_id")
    private DiningTable table;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "cashier_id")
    private UserInfm cashier;

    @Column(name = "guest_count")
    private Integer guestCount;

    @Builder.Default
    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<OrderItem> items = new ArrayList<>();

    @Builder.Default
    @Column(name = "subtotal", nullable = false, precision = 12, scale = 2)
    private BigDecimal subtotal = BigDecimal.ZERO;

    @Builder.Default
    @Column(name = "discount", nullable = false, precision = 12, scale = 2)
    private BigDecimal discount = BigDecimal.ZERO;

    @Builder.Default
    @Column(name = "vat_rate", nullable = false, precision = 5, scale = 2)
    private BigDecimal vatRate = BigDecimal.ZERO;

    @Builder.Default
    @Column(name = "vat_amount", nullable = false, precision = 12, scale = 2)
    private BigDecimal vatAmount = BigDecimal.ZERO;

    @Builder.Default
    @Column(name = "total", nullable = false, precision = 12, scale = 2)
    private BigDecimal total = BigDecimal.ZERO;

    @Builder.Default
    @Column(name = "total_khr", nullable = false, precision = 14, scale = 0)
    private BigDecimal totalKhr = BigDecimal.ZERO;

    @Enumerated(EnumType.STRING)
    @Column(name = "payment_method", length = 20)
    private PaymentMethod paymentMethod;

    @Column(name = "amount_tendered", precision = 12, scale = 2)
    private BigDecimal amountTendered;

    @Column(name = "change_amount", precision = 12, scale = 2)
    private BigDecimal changeAmount;

    @Builder.Default
    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 20)
    private OrderStatus status = OrderStatus.OPEN;

    @Column(name = "paid_at")
    private LocalDateTime paidAt;

    /* ---- Helpers that keep both sides of the association in sync --------- */

    public void addItem(OrderItem item) {
        items.add(item);
        item.setOrder(this);
    }

    public void removeItem(OrderItem item) {
        items.remove(item);
        item.setOrder(null);
    }

    public void clearItems() {
        items.forEach(i -> i.setOrder(null));
        items.clear();
    }
}

File: frontend/src/types/order.ts
export type OrderStatus = "OPEN" | "PAID" | "CANCELLED";
export type PaymentMethod = "CASH" | "CARD" | "KHQR" | "TRANSFER";

export interface OrderItem {
  id: number | null;
  productId: number | null;
  productName: string;
  qty: number;
  unitPrice: number;
  lineTotal: number;
  note: string | null;
}

export interface Order {
  id: number;
  invoiceNo: string;
  tableId: number | null;
  tableName: string | null;
  cashierId: number | null;
  cashierName: string | null;
  guestCount: number | null;
  items: OrderItem[];
  subtotal: number;
  discount: number;
  vatRate: number;
  vatAmount: number;
  total: number;
  totalKhr: number;
  paymentMethod: PaymentMethod | null;
  amountTendered: number | null;
  changeAmount: number | null;
  status: OrderStatus;
  createdAt: string | null;
  paidAt: string | null;
}

/** One line of the local basket, before it is sent to the server. */
export interface CartLine {
  productId: number;
  productName: string;
  unitPrice: number;
  qty: number;
  note?: string;
}

export const PAYMENT_ICON: Record<PaymentMethod, string> = {
  CASH: "💵",
  CARD: "💳",
  KHQR: "📱",
  TRANSFER: "🏦",
};

/** Response of {@code GET /api/orders/{id}/receipt}. */
export interface Receipt {
  restaurantName: string;
  restaurantNameEn: string;
  address: string;
  phone: string;
  order: Order;
}
